### PR TITLE
feat: added users api routes

### PR DESCRIPTION
**Pull request by willfpearce**
#72 
Created authenticated GET, DELETE, and PUT routes for "/api/users/:id"
I've tested all three with fetch and they work, but might need some tweaking

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -56,6 +57,61 @@ app.get(`/${config.admin_route}`, (req, res) => {
...
+  (req, res, next) => {
+    userModel.findById(req.params.id, (err, queryData) => {
+      if (!err) {
+        const userData = {
+          id: queryData._id,
+          username: queryData.username,
+          isAdmin: queryData.isAdmin,
+          date: queryData.date
+        }
+        res.status(200).send(userData)
```
Comment (zeim839) [inline app.js:74]: Change all route responses to
```js
return res.status(...).send(...)
```

**Review by zeim839 — CHANGES_REQUESTED** [review]
Could you create an additional GET route at `/api/users` that returns the first 20 users in the database?

File: app.js (hunk 1 above)
Comment (willfpearce) [reply app.js]: what does the return do?

File: app.js [hunk 2]
```diff
@@ -56,6 +57,102 @@ app.get(`/${config.admin_route}`, (req, res) => {
...
+
+// Users
+// gets and sends back first 20 users in db, starting with admins (sorting can be improved)
+app.get('/api/users/', passport.authenticate('loggedIn', { session: false }),
+  (req, res, next) => {
+    userModel.find({}, null, { limit: 20 }).sort({ isAdmin: -1 }).exec((err, docs) => {
+      if (!err) {
+        // remove password fields from docs
+        const responseDocs = docs.map(doc => {
+          return {
```
Comment (zeim839) [inline app.js:70]: You can remove the password field by doing:
```js
delete doc['password']
```

File: app.js [hunk 3]
```diff
@@ -56,6 +57,102 @@ app.get(`/${config.admin_route}`, (req, res) => {
...
+  }
+)
+
+// delete user and send user back
+app.delete('/api/users/:id',
+  passport.authenticate('loggedIn', { session: false }),
+  (req, res, next) => {
+    userModel.findByIdAndDelete(req.params.id, (err, doc) => {
+      if (!err) {
+        const userData = {
```
Comment (zeim839) [inline app.js:113]:
```js
delete doc['password']
return res.status(200).send(doc)
```

File: app.js (hunk 2 above)
Comment (willfpearce) [reply app.js:70]: fair enough

Comment (willfpearce) [reply app.js:70]: for some reason `delete doc.password` or `delete doc['password']` didn't work but `doc.password = undefined` does

Comment (willfpearce) [reply app.js:70]: @zeim839 I'm going to add some more sanity checking for the put method before I push

**Comment by willfpearce** [comment]
@zeim839 what do you think?

File: app.js [hunk 4]
```diff
@@ -56,6 +57,93 @@ app.get(`/${config.admin_route}`, (req, res) => {
...
+        return res.status(200).send(doc)
+      } else {
+        return res.status(404).send({ Error: 'User not found' })
+      }
+    })
+  }
+)
+
+// edit user and send updated user back
+app.put('/api/users/:id',
```
Comment (zeim839) [inline app.js:112]: Change the edit user route to:
```js
const mongoose = require('mongoose')
const { userModel } = require('./model/users')
const { hashPassword } = require('./auth/auth')
// ...

// edit user and send updated user back
app.put('/api/users/:id',
  passport.authenticate('loggedIn', { session: false }),
  (req, res, next) => {
    // sanity checking
    if (Object.keys(req.body).length === 0) {
      return res.status(400).send({ Error: 'Request body must include 1 or more property to update' })
    }

    if (req.body._id) {
      return res.status(400).send({ Error: 'Cannot change ID' })
    }

    if (!mongoose.Types.ObjectId.isValid(req.params.id)) {
      return res.status(400).send({ error: 'Invalid user id' })
    }

    if (req.body.password) {
      const { hash, salt } = hashPassword(req.body.password)
      req.body.password = { hash, salt }
    }

    const updatedUser = new userModel({ ...req.body })
    userModel.updateOne({ _id: req.params.id }, updatedUser._doc, async (err, doc) => {
      if (err || doc.modifiedCount <= 0) {
        return res.send(400).status({ error: 'Bad Request' })
      }

      const user = await userModel.findById(req.params.id)
      user.password = undefined
      return res.status(200).send(user)
    })
  }
)
```

**Review by zeim839 — CHANGES_REQUESTED** [review]
Could you also implement:
```js
app.get('/api/users',
  passport.authenticate('loggedIn', { session: false }),
  (req, res, next) => {
       // Return a list of all users
  }
)
```
And:
```js
app.post('/api/newuser',
  passport.authenticate('loggedIn', { session: false }),
  (req, res, next) => {
       // Create a new user
  }
)
```

**Comment by willfpearce** [comment]
> Could you also implement:
> 
> ```js
> app.get('/api/users',
>   passport.authenticate('loggedIn', { session: false }),
>   (req, res, next) => {
>        // Return a list of all users
>   }
> )
> ```
> 

@zeim839 how would we differentiate this from the route that sends back 20 users?

**Comment by zeim839** [comment]
> how would we differentiate this from the route that sends back 20 users?

You're right, you can ignore that. I was looking at the blog route earlier and got confused.

**Comment by willfpearce** [comment]
@zeim839 The put method you wrote only works if the req body has the complete user data. Is this intended?

**Review by zeim839 — APPROVED** [review]

**Pull request merged**